src/operator/tensor/dot-inl.h

```diff
@@ -216,32 +223,60 @@ inline bool DotForwardInferStorageType(const nnvm::NodeAttrs& attrs,
   bool only_lhs_transpose = param.transpose_a && !param.transpose_b;
   bool rhs_rsp_or_dns =
       rhs_stype == kRowSparseStorage || rhs_stype == kDefaultStorage;
+  NDArrayStorageType target_stype;
   if (!dispatched && lhs_stype == kDefaultStorage &&
       rhs_stype == kDefaultStorage) {
     // dns, dns -> dns
-    dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
-                                     DispatchMode::kFCompute);
+    target_stype = (param.forward_stype_hint.has_value())?
```

Comment:
nit: This line is quite long. Maybe cache the value of `param.forward_stype_hint.has_value()`?

Reply:
Done